transport/server.cc

```diff
@@ -620,11 +620,18 @@ future<> cql_server::connection::process_request() {
                             return _process_request_stage(this, istream, op, stream, service::client_state(service::client_state::request_copy_tag{}, client_state, ts), tracing_requested);
                         });
                     }
-                }().then_wrapped([this, buf = std::move(buf), mem_permit = std::move(mem_permit), leave = std::move(leave)] (future<processing_result> response_f) {
+                }().then_wrapped([this, buf = std::move(buf), mem_permit = std::move(mem_permit), leave = std::move(leave)] (future<processing_result> response_f) mutable {
                   try {
                     auto response = response_f.get0();
                     update_client_state(response);
-                    write_response(std::move(response.cql_response), _compression);
+                    // Inlined write_response call allows passing deferred gate leaving to its continuation
+                    _ready_to_respond = _ready_to_respond.then([this, response = std::move(response.cql_response), leave = std::move(leave)] () mutable {
+                        auto message = response->make_message(_version, _compression);
+                        message.on_delete([response = std::move(response)] { });
+                        return _write_buf.write(std::move(message)).then([this, leave = std::move(leave)] () mutable {
+                            return _write_buf.flush().finally([leave = std::move(leave)] {});
+                        });
+                    });
```

Comment:
Not thrilled about the code duplication.

How about renaming write_response to write_response_and_leave_gate() and adjusting all the call sites? That also ensures we haven't made the same mistake elsewhere.

The code here can do leave.cancel(); write_response_and_leave_gate(). This pair is awkward, as it the assymetric gate handling, but I think it's better than the duplication.

Anoterh option is some class that can be passed as a parameter to write_response instead of the anonymously typed leave.